### PR TITLE
[Android] Implement DiagnosticLog

### LINKED ISSUE #31647: [Android] Add diagnostic log capability in java controller and android chip-tool
Similar to https://github.com/project-chip/connectedhomeip/pull/31638, we need add the corresponding support in android chip tool

### PR DESCRIPTION
**Pull request by joonhaengHeo**
Fix #31647 

Implement DiagnosticLog in Android ChipTool and Java controller.

Test:
Add new DiagnosticLog test inside JavaMatterController in CI
Using real aqara device for testing.

### CODE REVIEW

File: src/protocols/bdx/BdxTransferServerDelegate.h [hunk 1]
```diff
@@ -46,7 +46,17 @@ class BDXTransferServerDelegate
      * @brief
      *  This method is invoked when a block is received.
      */
-    virtual CHIP_ERROR OnTransferData(BDXTransferProxy * transfer, const ByteSpan & data) = 0;
+    virtual CHIP_ERROR OnTransferData(BDXTransferProxy * transfer, const ByteSpan & data)
+    {
+        return CHIP_ERROR_NOT_IMPLEMENTED;
+    }
+
+    virtual CHIP_ERROR OnTransferData(BDXTransferProxy * transfer, const ByteSpan & data, bool isEof)
```
Comment (bzbarsky-apple) [inline src/protocols/bdx/BdxTransferServerDelegate.h:54]: Why is this needed?  The transfer ends OnTransferEnd, no?

Comment (joonhaengHeo) [reply src/protocols/bdx/BdxTransferServerDelegate.h:54]: I understand. I will fix it.

File: src/controller/java/BdxDiagnosticLogsReceiver.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+void BdxDiagnosticLogsReceiver::CancelBDXTransferTimeout()
+{
+    ChipLogProgress(Controller, "CancelBDXTransferTimeout");
+    chip::DeviceLayer::SystemLayer().CancelTimer(OnTransferTimeout, static_cast<void *>(this));
+}
+
+void BdxDiagnosticLogsReceiver::OnTransferTimeout(chip::System::Layer * layer, void * context)
+{
+    ChipLogProgress(Controller, "OnTransferTimeout");
+    auto * self = static_cast<BdxDiagnosticLogsReceiver *>(context);
```
Comment (yunhanw-google) [inline src/controller/java/BdxDiagnosticLogsReceiver.cpp:114]: please check if this is nullptr, thanks

File: src/controller/java/BdxDiagnosticLogsReceiver.cpp [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    mOnBdxTransferFailureCallback = onFailure;
+
+    mFabricIndex    = fabricIndex;
+    mNodeId         = nodeId;
+    mFileDesignator = fileDesignator;
+}
+
+CHIP_ERROR BdxDiagnosticLogsReceiver::OnTransferBegin(chip::bdx::BDXTransferProxy * transfer)
+{
+    chip::CharSpan fileDesignator = transfer->GetFileDesignator();
```
Comment (yunhanw-google) [inline src/controller/java/BdxDiagnosticLogsReceiver.cpp:45]: please check if transfer is nullptr or not, return the invalid argument if it is nullptr, thanks

File: src/controller/java/AndroidLogDownloadFromNode.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    chip::Callback::Callback<OnDeviceConnected> mOnDeviceConnectedCallback;
+    chip::Callback::Callback<OnDeviceConnectionFailure> mOnDeviceConnectionFailureCallback;
+    chip::Callback::Callback<OnBdxTransfer> mOnBdxTransferCallback;
+    chip::Callback::Callback<OnBdxTransferSuccess> mOnBdxTransferSuccessCallback;
+    chip::Callback::Callback<OnBdxTransferFailure> mOnBdxTransferFailureCallback;
+
+    chip::JniGlobalReference mJavaCallback;
+    NodeId mRemoteNodeId;
+    app::Clusters::DiagnosticLogs::IntentEnum mIntent;
+    uint16_t mTimeout;
```
Comment (yunhanw-google) [inline src/controller/java/AndroidLogDownloadFromNode.h:68]: please pre-initialize these member variables, thanks

File: src/controller/java/AndroidLogDownloadFromNode.cpp [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+    self->OnTransferCallback(fabricIndex, remoteNodeId, data, errInfoOnFailure);
+}
+
+void AndroidLogDownloadFromNode::OnTransferCallback(FabricIndex fabricIndex, NodeId remoteNodeId, const chip::ByteSpan & data,
+                                                    CHIP_ERROR * errInfoOnFailure)
+{
+    VerifyOrReturn(mJavaCallback.HasValidObjectRef(), ChipLogError(Controller, "mJavaCallback is invalid"));
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline src/controller/java/AndroidLogDownloadFromNode.cpp:252]: please check if env is nullptr or not, thanks

**Review by yunhanw-google — APPROVED** [review]

File: examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
+/*
+ *   Copyright (c) 2023 Project CHIP Authors
```
Comment (yunhanw-google) [inline examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt:2]: please update copy with 2024

File: examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import chip.devicecontroller.ChipDeviceController
+import chip.devicecontroller.DiagnosticLogType
+import chip.devicecontroller.DownloadLogCallback
+import com.matter.controller.commands.common.CredentialsIssuer
+import com.matter.controller.commands.common.MatterCommand
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicLong
+
+class DownloadLogCommand(controller: ChipDeviceController, credsIssuer: CredentialsIssuer?) :
```
Comment (yunhanw-google) [inline examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt:29]: could you add this DownloadLogCommand.kt test in
https://github.com/project-chip/connectedhomeip/blob/master/.github/workflows/java-tests.yaml#L239

**Review by yunhanw-google — APPROVED** [review]

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java [hunk 8]
```diff
@@ -883,6 +883,11 @@ public boolean openPairingWindowWithPINCallback(
         deviceControllerPtr, devicePtr, duration, iteration, discriminator, setupPinCode, callback);
   }
 
+  public boolean downloadLogFromNode(
```
Comment (yunhanw-google) [inline src/controller/java/src/chip/devicecontroller/ChipDeviceController.java:886]: could you add the doxygen for this public API? thanks

File: src/controller/java/src/chip/devicecontroller/DownloadLogCallback.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
+/*
+ *   Copyright (c) 2022 Project CHIP Authors
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/DownloadLogCallback.java:2]: Modified. Thanks.

File: src/controller/java/BdxDiagnosticLogsReceiver.cpp (hunk 3 above)
Comment (joonhaengHeo) [reply src/controller/java/BdxDiagnosticLogsReceiver.cpp:45]: Modified. Thanks.

File: src/controller/java/AndroidLogDownloadFromNode.h (hunk 4 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidLogDownloadFromNode.h:68]: Modified. Thanks.

File: src/controller/java/AndroidLogDownloadFromNode.cpp (hunk 5 above)
Comment (joonhaengHeo) [reply src/controller/java/AndroidLogDownloadFromNode.cpp:252]: Modified. Thanks.

File: examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt (hunk 6 above)
Comment (joonhaengHeo) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt]: Modified. Thanks.

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java (hunk 8 above)
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/ChipDeviceController.java:886]: Added. Thanks.

File: examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt (hunk 7 above)
Comment (joonhaengHeo) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/bdx/DownloadLogCommand.kt:29]: Added. Thanks.

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java [hunk 10]
```diff
@@ -883,6 +883,20 @@ public boolean openPairingWindowWithPINCallback(
...
+   * This function is used for downloading logs from the device.
+   *
+   * @param deviceId The 64-bit node ID of the device.
+   * @param type The log type being downloaded. See detailed in {@link DiagnosticLogType}.
+   * @param timeout This function sets the timeout. If set to 0, there will be no timeout.
+   * @param callback The callback is registered to convey the status during log downloads. See
+   *     detailed in {@link DownloadLogCallback}.
+   */
+  public boolean downloadLogFromNode(
+      long deviceId, DiagnosticLogType type, int timeout, DownloadLogCallback callback) {
```
Comment (yunhanw-google) [inline src/controller/java/src/chip/devicecontroller/ChipDeviceController.java:896]: int ==> long for timeout

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 11]
```diff
@@ -2059,6 +2060,26 @@ JNI_METHOD(jboolean, openPairingWindowWithPINCallback)
     return true;
 }
 
+JNI_METHOD(void, downloadLogFromNode)
+(JNIEnv * env, jobject self, jlong handle, jlong deviceId, jint typeEnum, jint timeout, jobject downloadLogCallback)
```
Comment (yunhanw-google) [inline src/controller/java/CHIPDeviceController-JNI.cpp:2064]: jlong timeout

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 12]
```diff
@@ -2059,6 +2060,26 @@ JNI_METHOD(jboolean, openPairingWindowWithPINCallback)
...
+{
+    chip::DeviceLayer::StackLock lock;
+    CHIP_ERROR err                           = CHIP_NO_ERROR;
+    AndroidDeviceControllerWrapper * wrapper = AndroidDeviceControllerWrapper::FromJNIHandle(handle);
```
Comment (yunhanw-google) [inline src/controller/java/CHIPDeviceController-JNI.cpp:2068]: check whether wrapper is nullptr or not

File: src/controller/java/CHIPDeviceController-JNI.cpp (hunk 11 above)
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp]: Fixed. Thanks.

File: src/controller/java/CHIPDeviceController-JNI.cpp (hunk 12 above)
Comment (joonhaengHeo) [reply src/controller/java/CHIPDeviceController-JNI.cpp:2068]: Fixed. Thanks.

File: src/controller/java/src/chip/devicecontroller/ChipDeviceController.java (hunk 10 above)
Comment (joonhaengHeo) [reply src/controller/java/src/chip/devicecontroller/ChipDeviceController.java]: Fixed. Thanks.

**Review by sharadb-amazon — APPROVED** [review]

File: src/controller/java/AndroidLogDownloadFromNode.cpp [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    VerifyOrReturnValue(logDownload != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    CHIP_ERROR err = logDownload->GetConnectedDevice();
+
+    if (err != CHIP_NO_ERROR)
+    {
+        delete logDownload;
+        logDownload = nullptr;
+    }
+    // Else will clean up when the callback is called.
```
[the remark this answers is not in the document]
Comment (joonhaengHeo) [reply src/controller/java/AndroidLogDownloadFromNode.cpp:82]: I undestand. Thank you for your comment.
